### ISSUE TITLE
UBsan: Hung check fails with `Connection reset by peer` due to `Memory limit (total) exceeded`

### ISSUE THREAD

**Issue by tavplubix** [ClickHouse/ClickHouse#45372]
https://s3.amazonaws.com/clickhouse-test-reports/44543/9ad6e1b1298782e5daf1ae3c05c448da7c4e5344/stress_test__ubsan_.html

```
2023-01-16 17:49:29,815 Checking if some queries hung
Using queries from '/usr/share/clickhouse-test/queries' directory
Using {args.client} as client program
Connecting to ClickHouse server...Connection error, will retry:  [Errno 104] Connection reset by peer
Connection error, will retry:  [Errno 104] Connection reset by peer
Connection error, will retry:  [Errno 104] Connection reset by peer
...
Connection error, will retry:  [Errno 104] Connection reset by peer
Connection error, will retry:  [Errno 104] Connection reset by peer
Connection error, will retry:  [Errno 104] Connection reset by peer

All connection tries failed
Server is not responding. Cannot execute 'SELECT 1' query.
Got server pid 1560

Located ClickHouse server process 1560 listening at TCP port 9000
Collecting stacktraces from all running threads with gdb:
...
```

Stacktraces: https://pastila.nl/?0006a09d/4680f83f4f1e0a34e304d9f70e78196b 
Postprocessed stacktraces: https://pastila.nl/?0006a09d/15ce6edb51b0502f7b1650cd6d5a985a

I did not find anything interesting in the stacktraces

Similar failure in master: https://s3.amazonaws.com/clickhouse-test-reports/0/4087ce4f71cdfb08644a87b114ead82dbb927508/stress_test__ubsan_.html

**Comment by tavplubix** [comment]
It's interesting that it always happens with ubsan

**Comment by tavplubix** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/45613/00c341850f6fce5dd8ee084d36b94afb99aa9be2/stress_test__ubsan_.html

**Comment by tavplubix** [comment]
The reason is:
```
$ rg -Faz "Memory limit (total) exceeded" clickhouse-server.stress.log.zst | wc
 436548 25903547 231686569
$ rg -Faz "Memory limit (total) exceeded" clickhouse-server.stress.log1.zst | tail -20
2023.01.16 18:00:26.346880 [ 1657 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1583891 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.355194 [ 1656 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1723447 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.359331 [ 1638 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1244763 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.360235 [ 1650 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1783675 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.363149 [ 1644 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1689415 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.365195 [ 1646 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1111125 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.376294 [ 1653 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1136112 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.381019 [ 1643 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1643777 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.382752 [ 1638 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1783675 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.393502 [ 1635 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1583891 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.401522 [ 1636 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 2088067 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.402047 [ 1653 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.62 GiB (attempt to allocate chunk of 1783675 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.404045 [ 1649 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.62 GiB (attempt to allocate chunk of 1052798 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.404126 [ 1634 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.62 GiB (attempt to allocate chunk of 1244763 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.409480 [ 1646 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.62 GiB (attempt to allocate chunk of 1111125 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.414790 [ 1641 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.62 GiB (attempt to allocate chunk of 1590373 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.422839 [ 1639 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.62 GiB (attempt to allocate chunk of 1126595 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.424540 [ 1636 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1783675 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.427252 [ 1645 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1583891 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.16 18:00:26.430191 [ 1650 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.63 GiB (attempt to allocate chunk of 1643777 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
```

It makes the server completely inoperable, so it cannot even accept connections. Need to find out who holds the memory. Maybe https://github.com/ClickHouse/ClickHouse/pull/46089 will help

**Comment by evillique** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/46420/eb3a7f55b933af87658b537505074c8b7091b358/stress_test__ubsan_.html

**Comment by SmitaRKulkarni** [comment]
Looks like the same issue https://s3.amazonaws.com/clickhouse-test-reports/46487/f3f459fe3411d7a86d2048455d502d134f57a01c/stress_test__ubsan_.html.
However  in the logs it says 'Cannot get server pid with pidof -s clickhouse-server' (https://s3.amazonaws.com/clickhouse-test-reports/46487/f3f459fe3411d7a86d2048455d502d134f57a01c/stress_test__ubsan_/hung_check.log)

**Comment by tavplubix** [comment]
> Looks like the same issue https://s3.amazonaws.com/clickhouse-test-reports/46487/f3f459fe3411d7a86d2048455d502d134f57a01c/stress_test__ubsan_.html. However in the logs it says 'Cannot get server pid with pidof -s clickhouse-server' (https://s3.amazonaws.com/clickhouse-test-reports/46487/f3f459fe3411d7a86d2048455d502d134f57a01c/stress_test__ubsan_/hung_check.log)

It's a different issue. "Cannot get server pid" means that the server process is not running for some reason. Usually it means that the server has crashed. And yes, there's also "Fatal message in clickhouse-server.log (see fatal_messages.txt)" and `Logical error: \'Cache usage is not allowed (query_id: 15764668-f9d5-44a9-bacd-67defd1d6ba7)\'.`

**Comment by SmitaRKulkarni** [comment]
> It's a different issue. "Cannot get server pid" means that the server process is not running for some reason. Usually it means that the server has crashed. And yes, there's also "Fatal message in clickhouse-server.log (see fatal_messages.txt)" and `Logical error: \'Cache usage is not allowed (query_id: 15764668-f9d5-44a9-bacd-67defd1d6ba7)\'.`

Thanks, will create a new issue

**Comment by alexey-milovidov** [comment]
This is not the case anymore.

**Comment by azat** [comment]
>It's interesting that it always happens with ubsan

TL;DR is here https://github.com/ClickHouse/ClickHouse/pull/76143
